### PR TITLE
Add deploy.sh for bento pipeline

### PR DESCRIPTION
**Pull request by hodgesto**
This is stage 5 of `!taste` and stage 1 of `!deploy`

### CODE REVIEW

File: files/default/bin/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+[ -z "$PR_NUM" ] && echo "Error: PR_NUM not set. Try '$0 -h'" && exit 1
+
+if [ $PUBLIC == 0 ]
+  then
+      [ -z $FILE ] && echo "Error: FILE not set. Try '$0 -h'" && exit 1
+      openstack image create --file $FILE --disk-format qcow2 "$PR_NUM"
+      exit 0
+else
+  [ -z "$IMG_NAME" ] && echo "Error: IMG_NAME not set. Try '$0 -h'" && exit 1
+  openstack image set --name "$IMG_NAME" --private "$PR_NUM"
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: The name would be `$IMG_NAME - $PR_NUM` because we might change multiple JSONS in one PR

File: files/default/bin/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  openstack image set --name "$PR_NUM" --public "$IMG_NAME"
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: Look above ^

**Comment by hodgesto** [comment]
@knightsamar Ready for Review.

File: files/default/bin/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  [ -z "$IMG_NAME" ] && echo "Error: IMG_NAME not set. Try '$0 -h'" && exit 1
+  [ -z "$PR_NUM" ] && echo "Error: PR_NUM not set. Try '$0 -h'" && exit 1
+
+if [ $PUBLIC == 0 ]
+  then
+    [ -z $FILE ] && echo "Error: FILE not set. Try '$0 -h'" && exit 1
+    openstack image create --file $FILE --disk-format qcow2 "$IMG_NAME - $PR_NUM"
+    exit 0
+else
+  OLD_ID=$(openstack image show "$IMG_NAME" | grep id | awk '{print $4}')
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: An easier way would be to `openstack image show "$IMG_NAME" -f value -c id`

File: files/default/bin/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  tput sgr0
+  echo "    -p          - Publish (publicly) the target image as '\$IMG_NAME', renaming the old"
+  echo "                  '\$IMG_NAME' to '\$IMG_NAME - \$PR_NUM' and making that image private"
+  echo
+  tput bold
+  echo " Arguments:"
+  tput sgr0
+  echo
+  echo "    FILE        - Fully built image to be uploaded to OpenStack if [-p] is"
+  echo "                  not specified. The image will be: '\$IMG_NAME - \$PR_NUM'"
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: "The image will be *named* as ... in OpenStack."

File: files/default/bin/deploy.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  echo
+  echo "    IMG_NAME    - Name of the image to replace if [-p] is specified."
+  echo
+  echo "    PR_NUM      - GitHub Pull Request number to name private images."
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: ...to suffix the private image names with.

File: files/default/bin/deploy.sh [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  echo
+  exit 0
+}
+
+while getopts "hpf:n:r:" opt ; do
+  case $opt in
+    h)
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: This should also be the default.

File: files/default/bin/deploy.sh [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    n)
+      IMG_NAME="$OPTARG"
+      ;;
+    r)
+      PR_NUM="$OPTARG"
+      ;;
+  esac
+done
+
+set -xe
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: Is this a debug leftover ? If not, you might want to move this after the error messages.

Comment (hodgesto) [inline files/default/bin/deploy.sh]: yep, sorry it was.

**Comment by hodgesto** [comment]
@knightsamar Ready for review

File: files/default/bin/deploy.sh [hunk 8]
```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+
+PUBLIC=0
+
+run_help() {
+  echo
+  tput bold
```
Comment (ramereth) [inline files/default/bin/deploy.sh]: Why is this needed?

**Comment by hodgesto** [comment]
by the way, should this be in https://github.com/osuosl-cookbooks/osl-jenkins/pull/75 ?

File: files/default/bin/deploy.sh (hunk 8 above)
Comment (hodgesto) [inline files/default/bin/deploy.sh]: Just for readability, it's not 100% necessary.

Comment (ramereth) [inline files/default/bin/deploy.sh]: IMO it's not needed and adds way too many extra lines to the file.

Comment (hodgesto) [inline files/default/bin/deploy.sh]: Ok, will take it out.

**Comment by knightsamar** [comment]
@hodgesto let's keep it separate from that PR for now. helps reviewing.

File: files/default/bin/deploy.sh [hunk 9]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
+PUBLIC=0
+
+run_help() {
+  echo
+  echo " NAME:\n"
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: The `\n` is unnecessary and for some reason seems to print as a literal for me.

Comment (hodgesto) [inline files/default/bin/deploy.sh]: Removed.

**Comment by hodgesto** [comment]
@knightsamar I have done a test run of this, both for building and publishing.

**Comment by hodgesto** [comment]
@knightsamar results of deploy.sh in use: https://pastebin.osuosl.org/60086/

File: files/default/bin/deploy.sh [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+if [ $PUBLIC == 0 ]
+  then
+    [ -z $FILE ] && echo "Error: FILE not set. Try '$0 -h'" && exit 1
+    openstack image create --file $FILE --disk-format qcow2 "$IMG_NAME - $PR_NUM"
+    exit 0
+else
+  OLD_ID=$(openstack image show "$IMG_NAME" -f value -c id)
+  openstack image set --name "$IMG_NAME" --public "$IMG_NAME - $PR_NUM"
+  openstack image set --name "$IMG_NAME - $PR_NUM" --private $OLD_ID
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: If this is the renaming of the old image which we was just deprecated by our newly deployed image, it should be suffixed as "deprecated by PR #".

File: files/default/bin/deploy.sh [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  echo
+  echo " OPTIONS:"
+  echo "    -p          - Publish (publicly) the target image as '\$IMG_NAME', renaming the old"
+  echo "                  '\$IMG_NAME' to '\$IMG_NAME - \$PR_NUM' and making that image private"
+  echo
+  echo " ARGUMENTS:"
+  echo "    FILE        - Fully built image to be uploaded to OpenStack if [-p] is not"
+  echo "                  specified. The image name will be: '\$IMG_NAME - \$PR_NUM'."
+  echo
+  echo "    IMG_NAME    - Prefix of the new image or name of the image to replace if"
```
Comment (knightsamar) [inline files/default/bin/deploy.sh]: How is this usable as a prefix?

File: files/default/bin/deploy.sh [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+[ -z $1 ] && run_help
+[ -z "$IMG_NAME" ] && echo "Error: IMG_NAME not set. Try '$0 -h'" && exit 1
+[ -z "$PR_NUM" ] && echo "Error: PR_NUM not set. Try '$0 -h'" && exit 1
+
+set -xe
+
+if [ $PUBLIC == 0 ]
+  then
+    [ -z $FILE ] && echo "Error: FILE not set. Try '$0 -h'" && exit 1
```
Comment (ramereth) [inline files/default/bin/deploy.sh]: I would change this to ``[ -r $FILE ]`` to verify that the file exists and is readable.

File: files/default/bin/deploy.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  exit 0
+}
+
+while getopts "hpf:n:r:" opt ; do
+  case ${opt} in
+    h)
+      run_help
+      ;;
+    p)
+      PUBLIC=1
```
Comment (ramereth) [inline files/default/bin/deploy.sh]: This variable is confusing as you describe it as "publish" in the help section and yet here you use "public". My vote is to keep it publish throughout so it's less confusing.

File: files/default/bin/deploy.sh (hunk 10 above)
Comment (ramereth) [inline files/default/bin/deploy.sh]: Otherwise it looks like you're giving it the same name as before and it would fail. I'd recommend you actually rename the old first, and then publish the new one otherwise you're going to have problems with conflicting names. The only other way around it is getting the UUID of both the new/old before you change and then reference those.

Actually, I think you should be using UUID's throughout so there are no problems. You can do that by running something like ``openstack image show "$IMG_NAME" -c id -f value``.

File: files/default/bin/deploy.sh (hunk 12 above)
Comment (hodgesto) [inline files/default/bin/deploy.sh]: -r gives me a false negative

Comment (hodgesto) [inline files/default/bin/deploy.sh]: I changed this but it is dependent on where the deploy.sh is in comparison to the image directory. I'm acting as if it will be placed in bin or something but I can change it as needed.

Comment (ramereth) [inline files/default/bin/deploy.sh]: Actually, it only matters where the command is being run *from*, not where the script is located.

**Comment by hodgesto** [comment]
@knightsamar @ramereth ready for review

File: files/default/bin/deploy.sh (hunk 12 above)
Comment (hodgesto) [inline files/default/bin/deploy.sh]: yes :) sorry

**Comment by hodgesto** [comment]
@ramereth as far as the `chef_version` goes, I will attempt to do that. But I think that's more of an image thing, correct? Or am I misunderstanding.

**Comment by ramereth** [comment]
@hodgesto yes, that's something you add to the ``openstack`` command when you upload the image.

**Comment by hodgesto** [comment]
@ramereth Ok, I'll mention https://github.com/osuosl/packer-templates/pull/16 to keep track over there, and I'll create a new PR to edit the CentOS images as well

**Comment by hodgesto** [comment]
@ramereth would you also like me to open a PR for the build script to put that in there?

**Comment by hodgesto** [comment]
@ramereth I realize this last commit is wrong, but I am trying to figure out how you meant for me to do this. there is no `--chef-version` option in `openstack image create`

File: files/default/bin/deploy.sh [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+[ -z $1 ] && run_help
+[ -z "$IMG_NAME" ] && echo "Error: IMG_NAME not set. Try '$0 -h'" && exit 1
+[ -z "$PR_NUM" ] && echo "Error: PR_NUM not set. Try '$0 -h'" && exit 1
+
+set -xe
+
+if [ $PUBLISH == 0 ]
+  then
+    [ -r ../$FILE ] && echo "Error: FILE not set. Try '$0 -h'" && exit 1
+    openstack image create --file $FILE --chef-version $CHEF_VER --disk-format qcow2 "$IMG_NAME - $PR_NUM"
```
Comment (ramereth) [inline files/default/bin/deploy.sh]: Oops, I missed that this is actually wrong. It should be: ``--property chef-version="$CHEF_VER"``

Comment (hodgesto) [inline files/default/bin/deploy.sh]: okay :) that's what I was about to change it to.

**Comment by knightsamar** [comment]
@hodgesto please check the script using `shellcheck` from your workstation.

**Comment by hodgesto** [comment]
@knightsamar fixed all issues given by `shellcheck`

**Comment by knightsamar** [comment]
@hodgesto please set the +x bit on the file, squash all commits and we are good to deploy the deploy!

**Comment by hodgesto** [comment]
@knightsamar squashed and ready to deploy

**Comment by knightsamar** [comment]
@hodgesto This should have gone as `!bump patch ~` :) When you are in next time, you would have to follow https://docs.osuosl.org/config-management/chef/cookbooks/updating-organization-cookbooks.html#the-old-way to actually deploy it.

**Comment by knightsamar** [comment]
I took care of this.
